src/table.c

```diff
@@ -0,0 +1,104 @@
+#include "vctrs.h"
+#include "slice.h"
+#include "utils.h"
+
+// Initialised at load time
+static SEXP syms_tbl_cast_dispatch = NULL;
+static SEXP fns_tbl_cast_dispatch = NULL;
+
+
+// Currently only data frames are considered tabular. Eventually
+// we should also allow arrays of dimensionality >= 2.
+
+// [[ include("vctrs.h") ]]
+bool vec_is_tabular(SEXP x) {
+  return is_data_frame(x);
+}
+
+// [[ include("vctrs.h") ]]
+void tbl_assert(SEXP x) {
+  if (!vec_is_tabular(x)) {
+    Rf_error("Input must be a data frame.");
+  }
+}
+// [[ register() ]]
+SEXP vctrs_tbl_assert(SEXP x) {
+  tbl_assert(x);
+  return R_NilValue;
+}
+// [[ register() ]]
+SEXP vctrs_tbl_is(SEXP x) {
+  return Rf_ScalarLogical(vec_is_tabular(x));
+}
+
+
+// [[ include("vctrs.h"); register() ]]
+SEXP tbl_slice(SEXP x, SEXP index) {
+  tbl_assert(x);
+  SEXP proxy = PROTECT(vec_proxy(x));
+
+  if (TYPEOF(proxy) != VECSXP) {
+    Rf_error("Internal error: Expected list in tabular proxy.");
+  }
+
+  SEXP names = PROTECT(r_names(x));
+  index = PROTECT(vec_as_index(index, vec_size(x), names));
```

Comment:
`vec_size(x) -> vec_size(proxy)`? i.e. we don't want the number of rows, we want the number of columns?

Reply:
This makes me wonder if we finally need a "getter" for "number of columns"

Reply:
oh yeah without this you can't reason about the final width after e.g. a cbind. I would call it `tbl_size()` though, unless we rename `vec_size()` to `vec_height()`.

Reply:
I thought about `tbl_size()` too, but I figured that a "table" still has a size that is the number of rows.

`vec_height()` feels wrong for a 1D vector like `1:5`.

One thought is to alias `vec_size()` as `tbl_height()` (maybe with a `tbl_assert()` check), and also have `tbl_width()`. Then you at least have a pair of functions that make sense together

Reply:
I think `vec_height()` makes sense for 1D vectors, because vctrs focuses on data frames.

> I figured that a "table" still has a size that is the number of rows.

When tables are conceived as vectors of rows, `vec_size()` is the relevant operator.

I don't really mind `tbl_width()`, though I prefer making the vec and tbl API symmetric.